### ISSUE TITLE
Not Enough Documentation

### ISSUE THREAD

**Issue by joesan** [docker/docker-bench-security#251]
I was looking around for some security stuff on Docker and I came across this. At first it sounds promising, but I'm totally unsure as to how I could run this. From the very little documentation that exists (in the README.md), I understand that this is that I should run this container on the host where my docker containers are running. I have my containers running inside a Kubernetest cluster and I have several of those Kubernetes clusters. So does this mean that I have to run one docker-bench-security container per Kubernetes cluster configuration?

Where could I find more information?

**Comment by r4j4h** [comment]
I agree, the demo docs are very biased towards having some folders that aren't present on many systems, but it does have text asking the user to look through that. It could be more friendly with more examples, the cross distro folder link is a nice nod towards that.

To your Kubernetes question, it's a very good one! I'm not sure if there is information for this particular use case, so here's some to get started:

You are exactly right in wanting to run this container on the host where the containers are running, so in Kubernetes that is on each `Node`. In particular, `DaemonSet`s are perfect for this, but they are designed to be long running. There are also `Job`s and `ScheduledJob`s which are meant for one-off tasks and for cron-like running of those one-off tasks. With `affinity` and `tolerance` settings you could probably get a job to cover each host, but that gets complicated.

I would go with the former and adapt the one-off run to a long-running container as described in [Run Once Daemonset On Kubernetes](http://blog.phymata.com/2017/05/13/run-once-daemonset-on-kubernetes/). Basically you run a DaemonSet with one container defined that runs this image with as many of the shared volume and privs as you can and with a modified command that runs the benchmark as normal and then echoes a "completed" signal and either sleep loops or tail -f /dev/null or some other method to hold the container open so Kubernetes doesn't start it up again. The idea is you have an outer process that checks them all for their "completed" signal and gathers their results and when finished deletes the daemonset (which should clean up the pods which should allow the image to be GC'd by Kubernetes, unless you run it again, etc).

**Comment by atolj** [comment]
Im standing for the same problem. I'm considering to execute the script on the nodes as describe by the following.

`git clone https://github.com/docker/docker-bench-security.git
cd docker-bench-security
sudo sh docker-bench-security.sh`

@r4j4h Is there a downside in using this approach?

**Comment by boeboe** [comment]
FYI,

This is the _daemonset_ configuration we use....

    apiVersion: extensions/v1beta1
    kind: DaemonSet
    metadata:
      name: docker-bench
      labels:
        name: docker-bench
    spec:
      updateStrategy:
        type: RollingUpdate
        rollingUpdate:
          maxUnavailable: 50%
      template:
        metadata:
          labels:
            name: docker-bench
        spec:
          serviceAccountName: cluster-reader
          hostPID: true
          hostIPC: true
          hostNetwork: true
          securityContext:
            runAsUser: 0
          containers:
            - name: docker-bench
              image: registry.bfed-pro.intapp.eu/docker/docker-bench-security:latest
              resources:
                requests:
                  cpu: 10m
                  memory: 32Mi
                limits:
                  cpu: 100m
                  memory: 64Mi
              securityContext:
                privileged: true
                capabilities:
                  add: ["AUDIT_CONTROL"]
              volumeMounts:
                - name: varlib
                  mountPath: /var/lib
                - name: varrun
                  mountPath: /var/run
                - name: usrlib
                  mountPath: /usr/lib
                - name: etc
                  mountPath: /etc
          volumes:
            - name: varlib
              hostPath:
                path: /var/lib
            - name: varrun
              hostPath:
                path: /var/run
            - name: usrlib
              hostPath:
                path: /usr/lib
            - name: etc
              hostPath:
                path: /etc
